### ISSUE TITLE
Exposing an automaton

### ISSUE THREAD

**Issue by hendrikvanantwerpen** [fancy-regex/fancy-regex#142]
👋🏻 Thanks for developing this crate!

I was wondering if there are any plans to expose underlying automatons? An example is the [regex-automata](https://crates.io/crates/regex-automata) crate that does this for `regex` supported regular expressions.

Our use case is that we are building up a text incrementally and want to keep track of regex matches while we are building it up. Re-executing the regex every time on the updated text would be too slow for our application.

**Comment by robinst** [comment]
There are no plans for that at the moment. Can you elaborate a bit more? What would the API look like?

**Comment by hendrikvanantwerpen** [comment]
The internals of the `regex` crate have such an interface: <https://docs.rs/regex-automata/latest/regex_automata/dfa/trait.Automaton.html>. The idea is that you have control over transitioning through the state. A sketch of a minimal set of operations is something like this:

    initial_state() -> State
    next_state(State) -> State
    is_match_state(State) -> bool
    is_dead_state(State) -> bool

One of the things this enables is incremental matching while building a string and for example aborting the string building if it doesn't match the regex (ends in a dead state).

I wonder if the interface from `regex-automata` is flexible enough for this crate. Probably not because the look-around is different, not restricted to a single byte.

**Comment by BurntSushi** [comment]
@hendrikvanantwerpen That's an API for one very specific type of regex engine: a pure DFA. If you look around the rest of `regex-automata`, you'll see other regex engines with different APIs. Consider the [`PikeVM`](https://docs.rs/regex-automata/latest/regex_automata/nfa/thompson/pikevm/struct.PikeVM.html) for example, which is much more coarsely grained and gives a lot less control. I would expect `fancy-regex`'s internals to look more like the latter than the former.

**Comment by hendrikvanantwerpen** [comment]
@BurntSushi Thanks for the pointers---I hadn't realized that the other flavors did not have similar automata defined.

To be clear, this was more a question out of interest than a strong suggestion :) I didn't think through the technical details at all, and certainly appreciate the effort it would take if the code isn't already structured like that under the hood.

The concrete problem that inspired this was solved differently in the end. Since the regexes only had a one-character lookahead, we used the `regex-automata` DFA with just a little bit of logic around it.
